### PR TITLE
Cfg stack

### PR DESCRIPTION
**Pull request by Christian-B**
fixes: https://github.com/SpiNNakerManchester/SpiNNUtils/issues/148

Should be done at the same time as:
https://github.com/SpiNNakerManchester/SpiNNMachine/pull/160
https://github.com/SpiNNakerManchester/SpiNNMan/pull/245
https://github.com/SpiNNakerManchester/PACMAN/pull/378
https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/800
https://github.com/SpiNNakerManchester/sPyNNaker/pull/1067
https://github.com/SpiNNakerManchester/SpiNNakerGraphFrontEnd/pull/190

DataSpecification not expected to ever needed a local cfg
spalloc has its won config system so specifically not inlcuded

These must be done at the same time due to the change on location of config_holder.py
https://github.com/SpiNNakerManchester/TestBase/pull/9
https://github.com/SpiNNakerManchester/SpiNNGym/pull/26

tested by ...

### CODE REVIEW

File: spinn_utilities/config_holder.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+                stripped = line.strip()
+                if len(stripped) == 0:
+                    continue
+                if stripped.startswith("#"):
+                    continue
+                if stripped.startswith("["):
+                    continue
+                if "Overwritten in Specific config file" in stripped:
+                    continue
+                option = stripped.split(" ")[0]
```
Comment (rowleya) [inline spinn_utilities/config_holder.py:281]: It is possible to read config files and iterate their contents.  Would this be better than manual parsing?

**Review by rowleya — CHANGES_REQUESTED** [review]
Looks OK - just one minor comment.

**Comment by rowleya** [comment]
All dependents are reviewed and approved.  Additional minor comments overall:
 - Would there be any sense in providing template configs at the level that the options are introduced?  Currently there is a template in FEC and sPyNNaker, but with options in them that are found deeper down the stack.
 - Is the plan that some of the deeper modules start to get config options from the config files?  Some provide config values but don't use them in the current code (I believe anyway)

File: spinn_utilities/config_holder.py (hunk 1 above)
Comment (Christian-B) [reply spinn_utilities/config_holder.py:281]: If you read the config in it silently overwrites what is found previously.

Which is great if you want the user cfg to overwrite the default and then the cfg in the script directory to overwrite that,

In this case I want to catch any cases where different layers of defaults have the same option.
Except in the specific cases where we expect overides

**Comment by Christian-B** [comment]
regarding templates.
I did wonder about that.
I am not sure the current code nicely handles templates where various cfgs have the same sections. ex [Mapping]

Will have a plan but probably a different PR

**Comment by Christian-B** [comment]
Q. Is the plan that some of the deeper modules start to get config options from the config files? Some provide config values but don't use them in the current code (I believe anyway)

A. Yes the whole purpose here is to have these values available earlier.
The values currently in are a subset of the ones I expect to move.

Once the change param_type to get_config pr is in the values should be in the first level that uses them.

File: spinn_utilities/config_holder.py (hunk 1 above)
Comment (rowleya) [reply spinn_utilities/config_holder.py:287]: That is true if you use the same ConfigParser, but you can create a new ConfigParser and call read on each run and it shouldn't overwrite anything.

However I now see that you are also looking for a specific comment, so yes, I guess this has to be a custom reader.

**Review by rowleya — APPROVED** [review]

**Pull request merged**